// @nx-go/nx-go/src/utils/normalize-options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeOptions = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nx/devkit");
const project_name_and_root_utils_1 = require("@nx/devkit/src/generators/project-name-and-root-utils");
const normalizeOptions = (tree, options, projectType, generator) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
    ensureProjectDirectory(options);
    const { projectName, projectRoot, projectNameAndRootFormat } = yield (0, project_name_and_root_utils_1.determineProjectNameAndRootOptions)(tree, {
        name: options.name,
        projectType: projectType,
        directory: options.directory,
        projectNameAndRootFormat: options.projectNameAndRootFormat,
        callingGenerator: generator,
    });
    const parsedTags = options.tags
        ? options.tags.split(',').map((s) => s.trim())
        : [];
    return Object.assign(Object.assign({}, options), { name: (0, devkit_1.names)(options.name).fileName, moduleName: (0, devkit_1.names)(projectName).propertyName.toLowerCase(), projectNameAndRootFormat,
        projectName,
        projectRoot,
        projectType,
        parsedTags });
});
exports.normalizeOptions = normalizeOptions;
const ensureProjectDirectory = (options) => {
    // Nx 20 BREAKING CHANGE: `directory` is now mandatory and `name` is optional
    if (devkit_1.NX_VERSION.startsWith('20') && options.directory === undefined) {
        options.directory = options.name;
    }
};
